<a href="https://colab.research.google.com/github/zzhining/ml_basic/blob/main/pandas_exercise_skeleton.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

np.random.seed(0)

# 과제 1
## Q1. 


<div class="alert alert-success">
[문제1]     
    
insurance.csv'파일 읽어서 df에 할당하고, 상위 5개 항목 출력하기
</div>

------------

[힌트]

pd.read_csv()

In [3]:
df = pd.read_csv('data/insurance.csv')
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

## Q2. 


<div class="alert alert-success">
[문제2]     
    
df의 요약(평균, 4분위수, 전체 갯수 등)정보 확인하기
</div>

------------

In [5]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

## Q3. 


<div class="alert alert-success">
[문제3]     
    
신규 컬럼(double bmi)을 생성하고, bmi의 2배가 되는 값을 할당하기
</div>

------------

In [7]:
df['double_bmi'] = df['bmi'] * 2
df

age     sex     bmi  children smoker     region      charges  double_bmi
0      19  female  27.900         0    yes  southwest  16884.92400       55.80
1      18    male  33.770         1     no  southeast   1725.55230       67.54
2      28    male  33.000         3     no  southeast   4449.46200       66.00
3      33    male  22.705         0     no  northwest  21984.47061       45.41
4      32    male  28.880         0     no  northwest   3866.85520       57.76
...   ...     ...     ...       ...    ...        ...          ...         ...
1333   50    male  30.970         3     no  northwest  10600.54830       61.94
1334   18  female  31.920         0     no  northeast   2205.98080       63.84
1335   18  female  36.850         0     no  southeast   1629.83350       73.70
1336   21  female  25.800         0     no  southwest   2007.94500       51.60
1337   61  female  29.070         0    yes  northwest  29141.36030       58.14

[1338 rows x 8 columns]

## Q4. 


<div class="alert alert-success">
[문제4]     
    
신규 컬럼(debt)을 생성하고, 0으로 채우기
</div>

------------

In [9]:
df['debt'] = 0
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     0  
1          67.54     0  
2          66.00     0  
3          45.41     0  
4          57.76     0  
...          ...   ...  
1333       61.94     0  
1334       63.84     0  
1335       73.70     0  
1336       51.60     0  
1337       58.14     0  

[1338 rows x 9 columns]

## Q5. 


<div class="alert alert-success">
[문제5]     
    
debt에 0부터 값이 증가하도록 채우기
</div>

------------

In [11]:
df['debt'] = range(0,1338)
df

age     sex     bmi  children smoker     region      charges  \
0      19  female  27.900         0    yes  southwest  16884.92400   
1      18    male  33.770         1     no  southeast   1725.55230   
2      28    male  33.000         3     no  southeast   4449.46200   
3      33    male  22.705         0     no  northwest  21984.47061   
4      32    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   50    male  30.970         3     no  northwest  10600.54830   
1334   18  female  31.920         0     no  northeast   2205.98080   
1335   18  female  36.850         0     no  southeast   1629.83350   
1336   21  female  25.800         0     no  southwest   2007.94500   
1337   61  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

## Q6. 


<div class="alert alert-success">
[문제6]     
    
1번 인덱스의 정보 출력
</div>

------------

[힌트]
iloc



age                  18  
sex                male  
bmi               33.77  
children              1  
smoker               no  
region        southeast  
charges         1725.55  
double bmi        67.54  
debt                  1  

In [12]:
df.iloc[1,:]

age                  18
sex                male
bmi               33.77
children              1
smoker               no
region        southeast
charges       1725.5523
double_bmi        67.54
debt                  1
Name: 1, dtype: object

## Q7. 


<div class="alert alert-success">
[문제7]     
    
'double bmi'열 삭제하기
</div>

In [14]:
df1 = df.drop('double_bmi', axis=1)
df1

age     sex     bmi  children smoker     region      charges  debt
0      19  female  27.900         0    yes  southwest  16884.92400     0
1      18    male  33.770         1     no  southeast   1725.55230     1
2      28    male  33.000         3     no  southeast   4449.46200     2
3      33    male  22.705         0     no  northwest  21984.47061     3
4      32    male  28.880         0     no  northwest   3866.85520     4
...   ...     ...     ...       ...    ...        ...          ...   ...
1333   50    male  30.970         3     no  northwest  10600.54830  1333
1334   18  female  31.920         0     no  northeast   2205.98080  1334
1335   18  female  36.850         0     no  southeast   1629.83350  1335
1336   21  female  25.800         0     no  southwest   2007.94500  1336
1337   61  female  29.070         0    yes  northwest  29141.36030  1337

[1338 rows x 8 columns]

## Q8. 


<div class="alert alert-success">
[문제8]     
    
 'age'열, 'sex'열만 출력하기
</div>

In [16]:
df.loc[:,'age':'sex']

age     sex
0      19  female
1      18    male
2      28    male
3      33    male
4      32    male
...   ...     ...
1333   50    male
1334   18  female
1335   18  female
1336   21  female
1337   61  female

[1338 rows x 2 columns]

# 과제2
## Q9.
<div class="alert alert-success">
[문제9]     
    
나이를 1살씩 증가시키기
</div>

In [18]:
df['age'] = df['age'] + 1

In [21]:
type(df['age'])

pandas.core.series.Series

In [19]:
df

age     sex     bmi  children smoker     region      charges  \
0      20  female  27.900         0    yes  southwest  16884.92400   
1      19    male  33.770         1     no  southeast   1725.55230   
2      29    male  33.000         3     no  southeast   4449.46200   
3      34    male  22.705         0     no  northwest  21984.47061   
4      33    male  28.880         0     no  northwest   3866.85520   
...   ...     ...     ...       ...    ...        ...          ...   
1333   51    male  30.970         3     no  northwest  10600.54830   
1334   19  female  31.920         0     no  northeast   2205.98080   
1335   19  female  36.850         0     no  southeast   1629.83350   
1336   22  female  25.800         0     no  southwest   2007.94500   
1337   62  female  29.070         0    yes  northwest  29141.36030   

      double_bmi  debt  
0          55.80     0  
1          67.54     1  
2          66.00     2  
3          45.41     3  
4          57.76     4  
...          ...   ...  
1333       61.94  1333  
1334       63.84  1334  
1335       73.70  1335  
1336       51.60  1336  
1337       58.14  1337  

[1338 rows x 9 columns]

## Q10.
<div class="alert alert-success">
[문제10]     
    
charges의 표시형식을 소수점 아래 2자리로 바꾸기
</div>

[힌트]

함수 생성, 적용(apply/map)

In [22]:
df['charges'] = df['charges'].round(2)

In [23]:
df

age     sex     bmi  children smoker     region   charges  double_bmi  \
0      20  female  27.900         0    yes  southwest  16884.92       55.80   
1      19    male  33.770         1     no  southeast   1725.55       67.54   
2      29    male  33.000         3     no  southeast   4449.46       66.00   
3      34    male  22.705         0     no  northwest  21984.47       45.41   
4      33    male  28.880         0     no  northwest   3866.86       57.76   
...   ...     ...     ...       ...    ...        ...       ...         ...   
1333   51    male  30.970         3     no  northwest  10600.55       61.94   
1334   19  female  31.920         0     no  northeast   2205.98       63.84   
1335   19  female  36.850         0     no  southeast   1629.83       73.70   
1336   22  female  25.800         0     no  southwest   2007.94       51.60   
1337   62  female  29.070         0    yes  northwest  29141.36       58.14   

      debt  
0        0  
1        1  
2        2  
3        3  
4        4  
...    ...  
1333  1333  
1334  1334  
1335  1335  
1336  1336  
1337  1337  

[1338 rows x 9 columns]

## Q11.
<div class="alert alert-success">
[문제11]     
    
컬럼 정렬(알파벳 순서대로) : age, bmi, charges... 순으로 정렬
</div>

[Hint]
sort_index()


In [24]:
df.sort_index(axis = 1)

age     bmi   charges  children  debt  double_bmi     region     sex  \
0      20  27.900  16884.92         0     0       55.80  southwest  female   
1      19  33.770   1725.55         1     1       67.54  southeast    male   
2      29  33.000   4449.46         3     2       66.00  southeast    male   
3      34  22.705  21984.47         0     3       45.41  northwest    male   
4      33  28.880   3866.86         0     4       57.76  northwest    male   
...   ...     ...       ...       ...   ...         ...        ...     ...   
1333   51  30.970  10600.55         3  1333       61.94  northwest    male   
1334   19  31.920   2205.98         0  1334       63.84  northeast  female   
1335   19  36.850   1629.83         0  1335       73.70  southeast  female   
1336   22  25.800   2007.94         0  1336       51.60  southwest  female   
1337   62  29.070  29141.36         0  1337       58.14  northwest  female   

     smoker  
0       yes  
1        no  
2        no  
3        no  
4        no  
...     ...  
1333     no  
1334     no  
1335     no  
1336     no  
1337    yes  

[1338 rows x 9 columns]

## Q12.
<div class="alert alert-success">
[문제12]     
    
값 정렬(age) : 나이 순으로 정렬
</div>

[Hint]
sort_values


In [37]:
df.sort_values(by= 'age', ascending=True)

age     sex     bmi  children smoker     region   charges  double_bmi  \
57     19    male  31.680         2    yes  southeast  34303.17       63.36   
31     19  female  26.315         0     no  northeast   2198.19       52.63   
1023   19    male  23.320         1     no  southeast   1711.03       46.64   
427    19  female  29.165         0     no  northeast   7323.73       58.33   
1317   19    male  53.130         0     no  southeast   1163.46      106.26   
...   ...     ...     ...       ...    ...        ...       ...         ...   
603    65  female  39.050         3     no  southeast  16085.13       78.10   
534    65    male  40.480         0     no  southeast  13831.12       80.96   
1241   65    male  36.960         2    yes  southeast  49577.66       73.92   
801    65  female  35.970         0     no  southeast  14313.85       71.94   
768    65  female  39.700         0     no  southwest  14319.03       79.40   

      debt  
57      57  
31      31  
1023  1023  
427    427  
1317  1317  
...    ...  
603    603  
534    534  
1241  1241  
801    801  
768    768  

[1338 rows x 9 columns]

## Q13.
<div class="alert alert-success">
[문제13]     
    
나이('age') 평균구하기
</div>

[Hint]
mean

In [6]:
df['age'].mean()

np.float64(39.20702541106129)

## Q14.
<div class="alert alert-success">
[문제14]     
    
보험료('charges')와 가장 상관관계가 높은 변수는?
</div>

[Hint]
 - 공분산: df.cov()
 - 상관계수: df.corr()

# 과제 3
## Q15.
<div class="alert alert-success">
[문제15]     
    
 남자, 여자가 각각 몇명있는지 구하기
</div>

[Hint]
pd.value_counts


## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가 3명인 데이터 출력하기
</div>

## Q16.
<div class="alert alert-success">
[문제16]     
    
자녀(children)가  5명 이상인 데이터 출력하기
</div>

## Q17.
<div class="alert alert-success">
[문제17]     
    
 bmi가  17 미만인 데이터 출력하기
</div>

# 과제 4 
## Q18.
<div class="alert alert-success">
[문제18]     
    
성별에 따른 보험료 평균
</div>

## Q19.
<div class="alert alert-success">
[문제19]     
    
흡연여부에 따른 보험료 평균
</div>

## Q20.
<div class="alert alert-success">
[문제20]     
    
지역별 보험료 평균
</div>

## Q21.
<div class="alert alert-success">
[문제21]     
    
흡연여부 + 지역별 데이터 평균
</div>

## Q22.
<div class="alert alert-success">
[문제22]     
    
 흡연여부 + 지역별 데이터 평균에서 'charges' 만 츨력하기
</div>

## Q23.
<div class="alert alert-success">
[문제23]     
    
 흡연여부 + 지역별 데이터 크기(size)
</div>

## Q24.
<div class="alert alert-success">
[문제24]     
    
smoker 여부에 따라, group_val 컬럼에 '흡연자', '비흡연자' 표시하기
</div>